src/GafferScene/PrimitiveSampler.cpp

```diff
@@ -0,0 +1,385 @@
+}
+
+bool PrimitiveSampler::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return
+		Deformer::affectsProcessedObject( input ) ||
+		input == sourcePlug() ||
+		input == sourceLocationPlug() ||
+		input == primitiveVariablesPlug() ||
+		input == prefixPlug() ||
+		input == statusPlug() ||
+		input == sourcePlug()->objectPlug() ||
+		input == inPlug()->transformPlug() ||
+		input == sourcePlug()->transformPlug() ||
+		affectsSamplingFunction( input )
+	;
+}
+
+void PrimitiveSampler::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	PrimitiveSampler::Deformer::hashProcessedObject( path, context, h );
+
+	const string sourceLocation = sourceLocationPlug()->getValue();
+	const string primitiveVariables = primitiveVariablesPlug()->getValue();
+	const string status = statusPlug()->getValue();
+	if( sourceLocation.empty() || ( primitiveVariables.empty() && status.empty() ) )
+	{
+		return;
+	}
+
+	ScenePlug::ScenePath sourcePath;
+	ScenePlug::stringToPath( sourceLocation, sourcePath );
+	if( !SceneAlgo::exists( sourcePlug(), sourcePath ) )
+	{
+		return;
+	}
+
+	h.append( sourcePlug()->objectHash( sourcePath ) );
+	h.append( primitiveVariables );
+	prefixPlug()->hash( h );
+	h.append( status );
+	h.append( inPlug()->fullTransformHash( path ) );
+	h.append( sourcePlug()->fullTransformHash( sourcePath ) );
+	hashSamplingFunction( h );
+}
+
+IECore::ConstObjectPtr PrimitiveSampler::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const Primitive *inputPrimitive = runTimeCast<const Primitive>( inputObject );
+	if( !inputPrimitive )
+	{
+		return inputObject;
+	}
+
+	const string sourceLocation = sourceLocationPlug()->getValue();
+	const string primitiveVariables = primitiveVariablesPlug()->getValue();
+	const string status = statusPlug()->getValue();
+	if( sourceLocation.empty() || ( primitiveVariables.empty() && status.empty() ) )
+	{
+		return inputObject;
+	}
+
+	PrimitiveVariable::Interpolation outputInterpolation = PrimitiveVariable::Invalid;
+	SamplingFunction samplingFunction = computeSamplingFunction( inputPrimitive, outputInterpolation );
+	if( !samplingFunction || outputInterpolation == PrimitiveVariable::Invalid )
+	{
+		return inputObject;
+	}
+
+	ScenePlug::ScenePath sourcePath;
+	ScenePlug::stringToPath( sourceLocation, sourcePath );
+	if( !SceneAlgo::exists( sourcePlug(), sourcePath ) )
+	{
+		return inputObject;
+	}
+
+	ConstObjectPtr sourceObject = sourcePlug()->object( sourcePath );
+	const Primitive *sourcePrimitive = runTimeCast<const Primitive>( sourceObject.get() );
+	if( !sourcePrimitive )
+	{
+		return inputObject;
+	}
+
+	ConstPrimitivePtr preprocessedSourcePrimitive = sourcePrimitive;
+	if( auto mesh = runTimeCast<const MeshPrimitive>( preprocessedSourcePrimitive.get() ) )
+	{
+		preprocessedSourcePrimitive = MeshAlgo::triangulate( mesh );
```

Comment:
Shame we have to do this per-location, how often would we have many locations sampling the same primitive?

Reply:
That's entirely possible, and it's not just triangulation we'd pay for repeatedly in that case, but also construction of the PrimitiveEvaluator (which does things like create acceleration structures). The problem is also somewhat general, in that things like Seeds nodes also create PrimitiveEvaluators internally. In the past we've done things like having global PrimitiveEvaluator caches, but I'm very reluctant to introduce another cache metric for people to have to tune. If anything, I'd consider a PrimitiveEvaluatorMakerNodeThingy that would output a PrimitiveEvaluator and therefore allow us to store it in Gaffer's main cache, but that requires rejigging bits of Cortex if it's to work for the Seeds node. For a first version, I think this is OK. Anecdotally, the IE equivalent of this node does not use the shared evaluator cache, even though it's located in the same library that implements that cache.